doc/release_notes.rst

```diff
@@ -57,6 +57,8 @@ E.g. if a new rule becomes available describe how to use it `make test` and in o
 
 * Adds CI to update keep pinned environment files up to date. `PR #1183 <https://github.com/pypsa-meets-earth/pypsa-earth/pull/1183>`__
 
+* Fix lossy bidirectional links, especially H2 pipelines, which would sometimes gain H2 instead of losing it.  `PR #1192 <https://github.com/pypsa-meets-earth/pypsa-earth/pull/1192>`__
+
```

Comment:
With the current barebone pipeline implementation, the H2 pipeline does not require any energy to operate (PR #1215 would change this). Instead, the pipeline receives an inflow of hydrogen at one bus (p0 if inflow at bus0 or p1 if inflow at bus1) and depending on the efficiency eta outputs are certain hydrogen flow at the other bus (p1 if inflow at bus0 or p0 if inflow at bus1).
p0 and p1 are linked as follows: p1 = -eta*p0
As a result the pipeline loses hydrogen substance when operating from bus0 to bus1, and gains hydrogen substance when operating from bus1 to bus0.
This fix addresses the latter as it is definitely problematic, that pipelines can generate hydrogen. Specifically, the fix splits the pipeline into two pipelines, that share the same p_nom_opt (ensured by a new constraint). Both pipelines can only operate from bus0 to bus1 and the new pipeline has bus0 and bus1 switched to enable flow from the original bus1 to the original bus0. To maintain the orginal pipeline cost, the capex of the new pipeline are set to 0.

Arguably, it is suboptimal as well, that the pipelines lose hydrogen substance instead of requiring electricity for compression. This would be addressed by PR #1215. With PR #1215 the pipeline behavior would look as follows:
p0 and p1 are still linked by p1 = -eta*p0, though eta is now calculated based on a static and a length dependend component.
In theory this means that H2 pipelines could still lose H2 substance, but in practice the efficiency values are setup in the config to result in eta=1 (same in PyPSA-Eur) so it doesn't lose any hydrogen. I think it's nice to still have option to considere pipeline leakage by changing the config.
In addition, the hydrogen inflow p0 and the electricity demand for compression p2 are linked via p2 = -eta2*p0, where eta2 is the compression energy per H2 inflow and is calculated based on the pipeline length.

Reply:
Ohh... Many thanks for the explanation. Haven't been aware of the details on how we are using electrical analogy (because `Link` component has been initially developed to represent a power flow). Frankly speaking, it sounds to me that it it would have more physical ground if we would treat `p` as pressure is using `Link` to represent the pipelines. Not sure however we can do it straight away 😄 

Thank you so much for explaining this!